### ISSUE TITLE
todo.txt Sorting and grouping in view mode

### ISSUE THREAD

**Issue by ramsan10** [gsantner/markor#859]
It is possible in markor to sort the entries in edit mode. This is not very convenient, as it may be desirable to view some ordering for the entries but keep the natural ordering in the file. This is convenient so as older entries are at the end of the file. One possible advantage is for using a version control system for the file, to compare with older files etc.

My proposal is to add a button to sort entries in view mode. Files does not change, only visualization. It should be possible to sort by project, date, priority, etc. (decreasing and increasing)

Also, it would be nice to group by projects.

**Comment by gsantner** [comment]
Feedback from me:

* View mode is really what is named for, View & readonly. It will stay readonly forever, I wont ever merge or add anything that does modifications in view mode.
* Grouping/Sorting/Searching in view mode: Unlimited possibilities, just needs somebody that does some css/js and send a Pull Request :-).

**Comment by gsantner** [comment]
@ramsan10 Any comment from you?

**Comment by ramsan10** [comment]
I agree that view mode should be readonly.

Grouping/Sorting/Searching in view mode: Unlimited possibilities. I also agree.

I would like to help on this, but I cannot, as I do not program css/js extensively.

**Comment by gsantner** [comment]
Adding to this "todo.txt view mode meta task":

*  Feature Request: Multi-level sort for todo #974 - I would rather suggest to add a proper ToDo "View mode" implementation and have these kind of stuff in represenation only. Where javascript etc can be used to browse todos more appealing. And in turn that view mode, can be like SimpleTasks and others... with fancy sections and stuff. Ehat SimpleTasks etc do is pretty much that, add a layer over the original data.

**Comment by the0s0ph1st** [comment]
The app has a Table of Contents option for Notes, can we have it for Tasks? The change needed to make it work is to tie Table of Contents with Sort. For example, if Sort by Context is selected, then opening Table of Contents will list the available contexts in todo.txt, and selecting a context will place the cursor at the first or last item with the context.

**Comment by tengucrow** [comment]
Maybe another suggestion - filters by context / projects / priorities. Nice realization in Simpletask - you can view/hide any tasks which is convenient when there are one or two hundreds of them.

**Comment by gsantner** [comment]
Yup. By the way edit mode has this implemented already, long press context/projrct/priority button.

**Comment by tengucrow** [comment]
Thanks. This works like markdown content - very useful. In the meantime,  while writing I realized that there is may be more effective method: not a single todo.txt, but rather bunch of *.todo.txt files splitted by contexts (look at image). The markor is perfect for it.
Now test drive it "in the field" :)
 
![Screenshot_2021-02-09-06-27-19-436_net gsantner markor](https://user-images.githubusercontent.com/17335453/107315944-5945f280-6aa0-11eb-8166-327888db3dfe.jpg)

**Comment by harshad1** [comment]
With the latest improvements to filtering by context and project, I feel that maybe we don't need a view mode for Todo.txt at all.

Instead, we should replace the view mode button with a filter button which will give the use the ability to do (one of)

1. Filter by context
2. Filter by project
3. Filter by due date
4. Filter by priority

All filtering will be done in search dialogs (which now support highlighting ) and in edit mode so the user can quickly select the filter results and perform an appropriate action (for example: mark as done)

What do you think?

**Comment by ramsan10** [comment]
The idea to filtering and ordering in view mode was that we wanted to view
the information in different ways without modifying the underlying txt
file. Is it possible to do it now with your proposal?

RR


Missatge de Harshad Vedartham ***@***.***> del dia dl., 19 de
jul. 2021 a les 8:18:

**Comment by harshad1** [comment]
To some degree. We have the ability to filter by context and project without modifying the underlying file right now. There will be some significant improvements to this in 2.8.

What I am proposing is that we extend this to due date and priority. The 'view' will be in a dialog similar to the search dialog.

Selecting multiple tasks for modification does involve modifying the underlying file in some case - if you select multiple non-contiguous tasks for editing, they will be grouped and moved to the top of the file for selection.

**Comment by gsantner** [comment]
> Instead, we should replace the view mode

I disagree, especially for proper export "view mode"/html is required.    
Where we develop more is not relevant to that though.

**Comment by harshad1** [comment]
> Where we develop more is not relevant to that though.

We could put all this functionality under search.

So search icon will open a dialog with the following options:

1. Search       <- Existing search dialog
2. Context      <- Existing search by context dialog
3. Project       <- Existing search by project dialog
4. Due date    <- New filter by due date dialog
5. Priority       <- New filter by priority dialog

Each will take you to the appropriate dialog

**Comment by harshad1** [comment]
Once the checkbox multi-select is merged, I can implement this pretty easily.

**Comment by harshad1** [comment]
@ramsan10

You can check out the filtering system here https://github.com/gsantner/markor/actions/runs/1108313641

Activate it by clicking on the title. Feedback would be appreciated

**Comment by harshad1** [comment]
@gsantner As a next stage for todo filtering, I would like to implement a simple query language. 

For example `(*A | *B) & #work` would mean select tasks with priority (A or B) and which have project #work. This would be loosely inspired by the [todoist filter language](https://todoist.com/help/articles/introduction-to-filters).

I want to do this in order to
1. Enable more complex filters
2. Allow filters to be saved as shortcuts in the 'Browse todo' dialog

any opinions / thoughts?

**Comment by gsantner** [comment]
I can only suggest to not reinvent what is pretty known out there, and java has good support for --- regex search

**Comment by harshad1** [comment]
My current thinking is a json structured language

```json
{
    "note": "Todo text filter in json format",
    "and": [
        "project": "work",
        "or" : [
            "context": "meeting",
            "context": "george"
        ],
        "not" : {
            "or": [
                "priority": "C",
                "priority": "D",
                "priority": "E"
            ]
        },
        "or": [
            "due": "today",
            "due": "overdue"
        ],
        "done": false
    ]
}
```

This will enable better filtering than a regex (must have correct project order etc in regex).

Each block is roughly like an s-expression. 

Since it is json, we will not need to write our own parser. Just a simple system to translate the json into a query (which is already handled by the todo txt filter dialog)

**Comment by gsantner** [comment]
but, who would spontaneously start typing such things..just to get some other representation of the text? It all looks more complex than it should be, or todo.txt itself is

**Comment by gsantner** [comment]
but, who would spontaneously start typing such things..just to get some other representation of the text? It all looks more complex than it should be, or todo.txt itself is.

The goal of this is to have a implementation of todo.txt for view mode. currently it's just the text shown "as is" in in html.

**Comment by harshad1** [comment]
The idea would be to have one or more saved filters. Then one click -> filtered tasks. What I have above is an example of what I would check in the morning before work.

It's harder to type than something like

```
(+work & ( @meeting | @george ) & !( *C | *D | *E ) & ( today | overdue ) & !done)
```

But with the latter we will have to write our own parser.

> The goal of this is to have a implementation of todo.txt for view mode. currently it's just the text shown "as is" in in html.

Yeah, this may not be the best place to have this discussion. I can move it to a discussion

**Comment by gsantner** [comment]
Yes ofc discussion is fine. But we don't need such a complex thing. Just all tasks with +work is fine. Or just all tasks with priority A.
